src/Riok.Mapperly/Helpers/MemberNamingUtil.cs

```diff
@@ -0,0 +1,123 @@
+namespace Riok.Mapperly.Helpers;
+
+public static class MemberNamingUtil
+{
+    private enum Naming
+    {
+        Lower,
+        Upper,
+    }
+
+    public static string ToCamelCase(this string str) => str.ToPascalCase().ToLowerFirstChar();
+
+    public static string ToPascalCase(this string str) =>
+        str.JoinUsingNaming("", wordsNaming: Naming.Lower, firstLetterNaming: Naming.Upper);
+
+    public static string ToSnakeCase(this string str) =>
+        str.JoinUsingNaming("_", wordsNaming: Naming.Lower, firstLetterNaming: Naming.Lower);
+
+    public static string ToUpperSnakeCase(this string str) => str.ToSnakeCase().ToUpperInvariant();
+
+    public static string ToKebabCase(this string str) =>
+        str.JoinUsingNaming("-", wordsNaming: Naming.Lower, firstLetterNaming: Naming.Lower);
+
+    public static string ToUpperKebabCase(this string str) => str.ToKebabCase().ToUpperInvariant();
+
+    private static string ToLowerFirstChar(this string str) => char.ToLowerInvariant(str[0]) + (str.Length is 1 ? "" : str[1..]);
+
+    private static string ToUpperFirstChar(this string str) => char.ToUpperInvariant(str[0]) + (str.Length is 1 ? "" : str[1..]);
+
+    private static string JoinUsingNaming(this string memberName, string separator, Naming wordsNaming, Naming firstLetterNaming)
+    {
+        if (string.IsNullOrWhiteSpace(memberName))
+        {
+            return memberName;
+        }
+
+        var words = new MemberNameToWordsSplitter(memberName).SplitIntoWords();
+        var wordsInCorrectNaming = wordsNaming switch
+        {
+            Naming.Lower => words.Select(w => w.ToLowerInvariant()),
+            Naming.Upper => words.Select(w => w.ToUpperInvariant()),
+            _ => throw new ArgumentOutOfRangeException($"{nameof(wordsNaming)} has an unknown value {wordsNaming}"),
+        };
+
+        var wordsWithCorrectFirstLetters = firstLetterNaming switch
+        {
+            Naming.Lower => wordsInCorrectNaming.Select(w => w.ToLowerFirstChar()),
+            Naming.Upper => wordsInCorrectNaming.Select(w => w.ToUpperFirstChar()),
+            _ => throw new ArgumentOutOfRangeException($"{nameof(firstLetterNaming)} has an unknown value {firstLetterNaming}"),
+        };
+
+        return string.Join(separator, wordsWithCorrectFirstLetters);
+    }
+
+    private class MemberNameToWordsSplitter(string memberName)
+    {
+        private string _acc = string.Empty;
```

Comment:
There are a lot of allocations happening here, I tried to optimize a bit and came up with the following solution. According to my (quickly done) benchmarks it should be a couple times faster with a lot less allocations and passes all tests. WDYT?

```c#
using System.Globalization;
using System.Text;

namespace Riok.Mapperly.Helpers;

public static class MemberNamingUtil
{
    public static string ToCamelCase(this string str) =>
        str.JoinWords("", LetterCase.Lower, LetterCase.Upper, LetterCase.Lower);

    public static string ToPascalCase(this string str) =>
        str.JoinWords("", LetterCase.Upper, LetterCase.Upper, LetterCase.Lower);

    public static string ToSnakeCase(this string str) =>
        str.JoinWords("_", LetterCase.Lower, LetterCase.Lower, LetterCase.Lower);

    public static string ToUpperSnakeCase(this string str)
        => str.JoinWords("_", LetterCase.Upper, LetterCase.Upper, LetterCase.Upper);

    public static string ToKebabCase(this string str) =>
        str.JoinWords("-", LetterCase.Lower, LetterCase.Lower, LetterCase.Lower);

    public static string ToUpperKebabCase(this string str)
        => str.JoinWords("-", LetterCase.Upper, LetterCase.Upper, LetterCase.Upper);

    private static string JoinWords(
        this string name,
        string separator,
        LetterCase startLetterCase,
        LetterCase wordStartLetterCase,
        LetterCase wordLetterCase)
    {
        if (name.Length == 0)
            return name;

        var capacity = name.Length;
        if (separator.Length > 0)
        {
            // average word length is around 5 chars
            // add 20% capacity to account for the separator
            capacity = (int)(capacity * 1.2);
        }

        var state = new ProcessorState(capacity, separator, startLetterCase, wordStartLetterCase, wordLetterCase);
        foreach (var c in name)
        {
            switch (char.GetUnicodeCategory(c))
            {
                case UnicodeCategory.UppercaseLetter:
                    state.AppendUpper(c);
                    break;
                case UnicodeCategory.LowercaseLetter:
                    state.AppendLower(c);
                    break;
                case UnicodeCategory.DecimalDigitNumber:
                    state.AppendDigit(c);
                    break;
                case UnicodeCategory.ConnectorPunctuation when !state.IsAtStart:
                    state.AppendSeparator();
                    break;
            }
        }

        return state.Done();
    }

    private struct ProcessorState(
        int capacity,
        string separator,
        LetterCase startLetterCase,
        LetterCase wordStartLetterCase,
        LetterCase wordLetterCase)
    {
        private readonly StringBuilder _sb = new(capacity);
        private readonly Func<char, char> _startMod = GetCharModifier(startLetterCase);
        private readonly Func<char, char> _wordStartMod = GetCharModifier(wordStartLetterCase);
        private readonly Func<char, char> _wordLetterMod = GetCharModifier(wordLetterCase);
        private State _state = State.Start;
        private State _prevState = State.Start;

        public bool IsAtStart => _state == State.Start;

        public string Done() => _sb.ToString();

        public void AppendSeparator() => UpdateState(State.Separator);

        public void AppendUpper(char c)
        {
            switch (_state)
            {
                case State.Start:
                    AppendStart(c, State.UppercaseLetter);
                    break;
                case State.LowercaseLetterOrDigit:
                case State.Separator:
                    AppendNewWord(c, State.UppercaseLetter);
                    break;
                case State.UppercaseLetter:
                    Append(c, State.UppercaseLetter);
                    break;
            }
        }

        public void AppendLower(char c)
        {
            switch (_state)
            {
                case State.Start:
                    AppendStart(c, State.LowercaseLetterOrDigit);
                    break;
                case State.Separator:
                    AppendNewWord(c, State.LowercaseLetterOrDigit);
                    break;
                case State.LowercaseLetterOrDigit:
                    Append(c, State.LowercaseLetterOrDigit);
                    break;
                case State.UppercaseLetter:
                    if (_prevState == State.UppercaseLetter)
                    {
                        var prevC = _sb[^1];
                        _sb.Remove(_sb.Length - 1, 1);
                        _sb.Append(separator);
                        _sb.Append(_wordStartMod(prevC));
                    }

                    Append(c, State.LowercaseLetterOrDigit);
                    break;
            }
        }

        public void AppendDigit(char c)
        {
            switch (_state)
            {
                case State.Start:
                    AppendStart(c, State.LowercaseLetterOrDigit);
                    break;
                case State.Separator:
                    AppendNewWord(c, State.LowercaseLetterOrDigit);
                    break;
                case State.LowercaseLetterOrDigit:
                case State.UppercaseLetter:
                    Append(c, State.LowercaseLetterOrDigit);
                    break;
            }
        }

        private void AppendNewWord(char c, State newState)
        {
            _sb.Append(separator);
            _sb.Append(_wordStartMod(c));
            UpdateState(newState);
        }

        private void AppendStart(char c, State newState)
        {
            _sb.Append(_startMod(c));
            UpdateState(newState);
        }

        private void Append(char c, State newState)
        {
            _sb.Append(_wordLetterMod(c));
            UpdateState(newState);
        }

        private void UpdateState(State newState)
        {
            _prevState = _state;
            _state = newState;
        }

        private static Func<char, char> GetCharModifier(LetterCase letterCase)
            => letterCase switch
            {
                LetterCase.Lower => char.ToLowerInvariant,
                LetterCase.Upper => char.ToUpperInvariant,
                _ => throw new ArgumentOutOfRangeException(nameof(letterCase), letterCase, "Unknown letter case")
            };

        private enum State
        {
            Start,
            UppercaseLetter,
            LowercaseLetterOrDigit,
            Separator,
        }
    }

    private enum LetterCase
    {
        Lower,
        Upper,
    }
}

```

Reply:
Wow! Your solution is really impressive. I've updated my implementation to use your approach.